### ISSUE TITLE
AWSIoTDataManager.getShadow() timeout [race condition with AWSIoTDataManager.register()?]

### ISSUE THREAD

**Issue by armandr** [aws-amplify/aws-sdk-ios#1159]
**State your question**

How does one know AWSIoTDataManager.register(withShadow) has **completed the registration**(finished subscribing)?

In my IoS app I'm calling AWSIoTDataManager.register() with QoS:1 and on success immediately calling AWSIoTDataManager.getShadow().  This often, but not always leads to a AWSIoTShadowOperationStatusTypeTimeout.

It seems possible for getShadow() to be called **before** register() task has completed subscribing to shadow topics therefore resulting in a timeout.

However, adding a slight delay (as low as 100ms, but more than 10us) in between the register() and getshadow() makes the result more deterministic and successful (at least on my machine and simulator)

 The BOOL return value of AWSIoTDataManager.register() is based on the input parameters and does not provide an ACK Callback.

How does one determine that the register is successful and complete so shadow get/update operations can begin?

**Which AWS Services are you utilizing?**
AWSIoT, AWSCognito

**Provide code snippets (if applicable)**
 ```
let res = self.iotDataManager.register(withShadow: "DATHING", options: ["shadowOperationTimeoutSeconds": "QoS": 1, ": true],
                        eventCallback: { (string1, optype, status, string2, data) in
                       // handle callback
                    })
                    
DispatchQueue.main.asyncAfter(deadline: .now()+0.5 ) {                        
      self.iotDataManager.getShadow("HUMTHING")
}

```
**Environment(please complete the following information):**
 - SDK Version: latest
 - Dependency Manager: Cocoapods
 - Swift Version : 4.2

**Device Information (please complete the following information):**
 - Device: Simulator
 - iOS Version: iOS latest
 - Specific to simulators:

**Comment by minbi** [comment]
Hi @armandr ,

I took a look at the code and it looks like we don't currently support passing this information back. The subscribe code returns void [on this line](https://github.com/aws-amplify/aws-sdk-ios/blob/master/AWSIoT/Internal/AWSIoTMQTTClient.m#L775). I'm marking this as a feature request and also assigning @cbommas who might have more/better information.

**Comment by AndreaPetrelli** [comment]
I'm in the same situation, there is any news?

**Comment by dholdren** [comment]
I'm also running into this issue. seems like a common use-case, register for updates and immediately get the device shadow

**Comment by kshrikant** [comment]
@dholdren  @AndreaPetrelli @minbi @armandr May i know how did you guys resolved this, I'm facing similar problem while getting shadow ?

**Comment by armandr** [comment]
@kshrikant  You can try adding a short delay (e.g. 500ms) before getting the shadow and repeat re-try if you still get a timeout.

It's not optimal, but will make the behaviour more deterministic.

**Comment by kshrikant** [comment]
@armandr Thank i've added currently delay, let's see if any problem occurs.

**Comment by ruisebas** [comment]
This has been released in [2.36.4](https://github.com/aws-amplify/aws-sdk-ios/releases/tag/2.36.4)
